File: 1833-maximum-ice-cream-bars/1833-maximum-ice-cream-bars.cpp
class Solution {
public:
    int maxIceCream(vector<int>& costs, int coins) {
        int maxCost = 0;
        for (int cost : costs) {
            maxCost = max(maxCost, cost);
        }
        
        vector<int> count(maxCost + 1, 0);
        for (int cost : costs) {
            count[cost]++;
        }
        
        int bars = 0;
        for (int cost = 1; cost <= maxCost && coins > 0; cost++) {
            if (count[cost] > 0) {
                int canBuy = min(count[cost], coins / cost);
                bars += canBuy;
                coins -= canBuy * cost;
            }
        }
        
        return bars;
    }
};

File: 1833-maximum-ice-cream-bars/1833-maximum-ice-cream-bars.py
from typing import List

class Solution:
    def maxIceCream(self, costs: List[int], coins: int) -> int:
        max_cost = max(costs)
        freq = [0] * (max_cost + 1)
        
        for c in costs:
            freq[c] += 1
        
        count = 0
        for cost in range(1, max_cost + 1):
            while freq[cost] > 0 and coins >= cost:
                coins -= cost
                count += 1
                freq[cost] -= 1
        
        return count